README.md
Arduino-DS3231
==============

DS3231 Real-Time Clock.

Date formats (Day)
------------------

 * d : Day of the month, 2 digits with leading zeros (01 to 31)
 * D : A textual representation of a day, three letters (Mon through Sun)
 * j : Day of the month without leading zeros (1 to 31)
 * l : A full textual representation of the day of the week (Sunday through Saturday)
 * N : ISO-8601 numeric representation of the day of the week (1 for Monday through 7 for Sunday)
 * S : English ordinal suffix for the day of the month, 2 characters (st, nd, rd or th. Works well with j)
 * w : Numeric representation of the day of the week (0 for Sunday through 6 for Saturday)
 * z : The day of the year (0 through 365)

Date formats (Month)
--------------------

 * F : A full textual representation of a month, such as January or March (January through December)
 * m : Numeric representation of a month, with leading zeros (01 through 12)
 * M : A short textual representation of a month, three letters (Jan through Dec)
 * n : Numeric representation of a month, without leading zeros (1 through 12)
 * t : Number of days in the given month (28 through 31)

Date formats (Year)
-------------------

 * L : Whether it's a leap year (1 if it is a leap year, 0 otherwise)
 * Y : A full numeric representation of a year, 4 digits (Examples: 1999 or 2003)
 * y : A two digit representation of a year (Examples: 99 or 03)

Date formats (Time)
-------------------

 * a : Lowercase Ante meridiem and Post meridiem (am or pm)
 * A : Uppercase Ante meridiem and Post meridiem (AM or PM)
 * g : 2-hour format of an hour without leading zeros (1 through 12)
 * G : 24-hour format of an hour without leading zeros (0 through 23)
 * h : 12-hour format of an hour with leading zeros (01 through 12)
 * H : 24-hour format of an hour with leading zeros (00 through 23)
 * i : Minutes with leading zeros (00 to 59)
 * s : Seconds, with leading zeros (00 through 59)

Dare formats (Full Date/Time)
-----------------------------

 * U : Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)

Credits
-------

Forked from https://github.com/jarzebski/Arduino-DS3231


CHANGELOG
DS3231 Arduino Library 1.0 / 03.02.2021
=========================================

 * First release



DS3231.h
/*
 *  DS3231.h
 *
 *  Header file for the DS3231 Real-Time Clock
 *
 *  created 03 Feb 2021
 *  by Lars Erik Storbukås <https://github.com/storbukas>
 *
 *  Source: https://github.com/storbukas/DS3231
 *  Original source: https://github.com/jarzebski/Arduino-DS3231>
 *
 *  Licensed under the GPL-3.0 License
 *  https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 */

#ifndef DS3231_h
#define DS3231_h

#if ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif

#define DS3231_ADDRESS (0x68)

#define DS3231_REG_TIME (0x00)
#define DS3231_REG_ALARM_1 (0x07)
#define DS3231_REG_ALARM_2 (0x0B)
#define DS3231_REG_CONTROL (0x0E)
#define DS3231_REG_STATUS (0x0F)
#define DS3231_REG_TEMPERATURE (0x11)

#ifndef RTCDATETIME_STRUCT_H
#define RTCDATETIME_STRUCT_H
struct RTCDateTime {
  uint16_t year;
  uint8_t month;
  uint8_t day;
  uint8_t hour;
  uint8_t minute;
  uint8_t second;
  uint8_t dayOfWeek;
  uint32_t unixtime;
};

struct RTCAlarmTime {
  uint8_t day;
  uint8_t hour;
  uint8_t minute;
  uint8_t second;
};
#endif

typedef enum {
  DS3231_1HZ = 0x00,
  DS3231_4096HZ = 0x01,
  DS3231_8192HZ = 0x02,
  DS3231_32768HZ = 0x03
} DS3231_sqw_t;

typedef enum {
  DS3231_EVERY_SECOND = 0b00001111,
  DS3231_MATCH_S = 0b00001110,
  DS3231_MATCH_M_S = 0b00001100,
  DS3231_MATCH_H_M_S = 0b00001000,
  DS3231_MATCH_DT_H_M_S = 0b00000000,
  DS3231_MATCH_DY_H_M_S = 0b00010000
} DS3231_alarm1_t;

typedef enum {
  DS3231_EVERY_MINUTE = 0b00001110,
  DS3231_MATCH_M = 0b00001100,
  DS3231_MATCH_H_M = 0b00001000,
  DS3231_MATCH_DT_H_M = 0b00000000,
  DS3231_MATCH_DY_H_M = 0b00010000
} DS3231_alarm2_t;

class DS3231 {
 public:
  bool begin(void);

  void setDateTime(uint16_t year, uint8_t month, uint8_t day, uint8_t hour,
                   uint8_t minute, uint8_t second);
  void setDateTime(uint32_t t);
  void setDateTime(const char* date, const char* time);
  RTCDateTime getDateTime(void);
  uint8_t isReady(void);

  DS3231_sqw_t getOutput(void);
  void setOutput(DS3231_sqw_t mode);
  void enableOutput(bool enabled);
  bool isOutput(void);
  void enable32kHz(bool enabled);
  bool is32kHz(void);

  void forceConversion(void);
  float readTemperature(void);

  void setAlarm1(uint8_t dydw, uint8_t hour, uint8_t minute, uint8_t second,
                 DS3231_alarm1_t mode, bool armed = true);
  RTCAlarmTime getAlarm1(void);
  DS3231_alarm1_t getAlarmType1(void);
  bool isAlarm1(bool clear = true);
  void armAlarm1(bool armed);
  bool isArmed1(void);
  void clearAlarm1(void);

  void setAlarm2(uint8_t dydw, uint8_t hour, uint8_t minute,
                 DS3231_alarm2_t mode, bool armed = true);
  RTCAlarmTime getAlarm2(void);
  DS3231_alarm2_t getAlarmType2(void);
  bool isAlarm2(bool clear = true);
  void armAlarm2(bool armed);
  bool isArmed2(void);
  void clearAlarm2(void);

  void setBattery(bool timeBattery, bool squareBattery);

  char* dateFormat(const char* dateFormat, RTCDateTime dt);
  char* dateFormat(const char* dateFormat, RTCAlarmTime dt);

 private:
  RTCDateTime t;

  char* strDayOfWeek(uint8_t dayOfWeek);
  char* strMonth(uint8_t month);
  char* strAmPm(uint8_t hour, bool uppercase);
  char* strDaySufix(uint8_t day);

  uint8_t hour12(uint8_t hour24);
  uint8_t bcd2dec(uint8_t bcd);
  uint8_t dec2bcd(uint8_t dec);

  long time2long(uint16_t days, uint8_t hours, uint8_t minutes,
                 uint8_t seconds);
  uint16_t date2days(uint16_t year, uint8_t month, uint8_t day);
  uint8_t daysInMonth(uint16_t year, uint8_t month);
  uint16_t dayInYear(uint16_t year, uint8_t month, uint8_t day);
  bool isLeapYear(uint16_t year);
  uint8_t dow(uint16_t y, uint8_t m, uint8_t d);

  uint32_t unixtime(void);
  uint8_t conv2d(const char* p);

  void writeRegister8(uint8_t reg, uint8_t value);
  uint8_t readRegister8(uint8_t reg);
};

#endif

DS3231.cpp
/*
 *  DS3231.cpp
 *
 *  Class file for the DS3231 Real-Time Clock
 *
 *  created 03 Feb 2021
 *  by Lars Erik Storbukås <https://github.com/storbukas>
 *
 *  Source: https://github.com/storbukas/DS3231
 *  Original source: https://github.com/jarzebski/Arduino-DS3231>
 *
 *  Licensed under the GPL-3.0 License
 *  https://www.gnu.org/licenses/gpl-3.0.en.html
 *
 */

#if ARDUINO >= 100
#include "Arduino.h"
#else
#include "WProgram.h"
#endif

#include <Wire.h>

#include "DS3231.h"

const uint8_t daysArray[] PROGMEM = {31, 28, 31, 30, 31, 30,
                                     31, 31, 30, 31, 30, 31};
const uint8_t dowArray[] PROGMEM = {0, 3, 2, 5, 0, 3, 5, 1, 4, 6, 2, 4};

bool DS3231::begin(void) {
  Wire.begin();

  setBattery(true, false);

  t.year = 2000;
  t.month = 1;
  t.day = 1;
  t.hour = 0;
  t.minute = 0;
  t.second = 0;
  t.dayOfWeek = 6;
  t.unixtime = 946681200;

  return true;
}

void DS3231::setDateTime(uint16_t year, uint8_t month, uint8_t day,
                         uint8_t hour, uint8_t minute, uint8_t second) {
  Wire.beginTransmission(DS3231_ADDRESS);

#if ARDUINO >= 100
  Wire.write(DS3231_REG_TIME);
#else
  Wire.send(DS3231_REG_TIME);
#endif

#if ARDUINO >= 100
  Wire.write(dec2bcd(second));
  Wire.write(dec2bcd(minute));
  Wire.write(dec2bcd(hour));
  Wire.write(dec2bcd(dow(year, month, day)));
  Wire.write(dec2bcd(day));
  Wire.write(dec2bcd(month));
  Wire.write(dec2bcd(year - 2000));
#else
  Wire.send(dec2bcd(second));
  Wire.send(dec2bcd(minute));
  Wire.send(dec2bcd(hour));
  Wire.send(dec2bcd(dow(year, month, day)));
  Wire.send(dec2bcd(day));
  Wire.send(dec2bcd(month));
  Wire.send(dec2bcd(year - 2000));
#endif

#if ARDUINO >= 100
  Wire.write(DS3231_REG_TIME);
#else
  Wire.send(DS3231_REG_TIME);
#endif

  Wire.endTransmission();
}

void DS3231::setDateTime(uint32_t t) {
  t -= 946681200;

  uint16_t year;
  uint8_t month;
  uint8_t day;
  uint8_t hour;
  uint8_t minute;
  uint8_t second;

  second = t % 60;
  t /= 60;

  minute = t % 60;
  t /= 60;

  hour = t % 24;
  uint16_t days = t / 24;
  uint8_t leap;

  for (year = 0;; ++year) {
    leap = year % 4 == 0;
    if (days < 365 + leap) {
      break;
    }
    days -= 365 + leap;
  }

  for (month = 1;; ++month) {
    uint8_t daysPerMonth = pgm_read_byte(daysArray + month - 1);

    if (leap && month == 2) {
      ++daysPerMonth;
    }

    if (days < daysPerMonth) {
      break;
    }
    days -= daysPerMonth;
  }

  day = days + 1;

  setDateTime(year + 2000, month, day, hour, minute, second);
}

void DS3231::setDateTime(const char *date, const char *time) {
  uint16_t year;
  uint8_t month;
  uint8_t day;
  uint8_t hour;
  uint8_t minute;
  uint8_t second;

  year = conv2d(date + 9);

  switch (date[0]) {
    case 'J':
      month = date[1] == 'a' ? 1 : month = date[2] == 'n' ? 6 : 7;
      break;
    case 'F':
      month = 2;
      break;
    case 'A':
      month = date[2] == 'r' ? 4 : 8;
      break;
    case 'M':
      month = date[2] == 'r' ? 3 : 5;
      break;
    case 'S':
      month = 9;
      break;
    case 'O':
      month = 10;
      break;
    case 'N':
      month = 11;
      break;
    case 'D':
      month = 12;
      break;
  }

  day = conv2d(date + 4);
  hour = conv2d(time);
  minute = conv2d(time + 3);
  second = conv2d(time + 6);

  setDateTime(year + 2000, month, day, hour, minute, second);
}

char *DS3231::dateFormat(const char *dateFormat, RTCDateTime dt) {
  char buffer[255];

  buffer[0] = 0;

  char helper[11];

  while (*dateFormat != '\0') {
    switch (dateFormat[0]) {
      // Day decoder
      case 'd':
        sprintf(helper, "%02d", dt.day);
        strcat(buffer, (const char *)helper);
        break;
      case 'j':
        sprintf(helper, "%d", dt.day);
        strcat(buffer, (const char *)helper);
        break;
      case 'l':
        strcat(buffer, (const char *)strDayOfWeek(dt.dayOfWeek));
        break;
      case 'D':
        strncat(buffer, strDayOfWeek(dt.dayOfWeek), 3);
        break;
      case 'N':
        sprintf(helper, "%d", dt.dayOfWeek);
        strcat(buffer, (const char *)helper);
        break;
      case 'w':
        sprintf(helper, "%d", (dt.dayOfWeek + 7) % 7);
        strcat(buffer, (const char *)helper);
        break;
      case 'z':
        sprintf(helper, "%d", dayInYear(dt.year, dt.month, dt.day));
        strcat(buffer, (const char *)helper);
        break;
      case 'S':
        strcat(buffer, (const char *)strDaySufix(dt.day));
        break;

      // Month decoder
      case 'm':
        sprintf(helper, "%02d", dt.month);
        strcat(buffer, (const char *)helper);
        break;
      case 'n':
        sprintf(helper, "%d", dt.month);
        strcat(buffer, (const char *)helper);
        break;
      case 'F':
        strcat(buffer, (const char *)strMonth(dt.month));
        break;
      case 'M':
        strncat(buffer, (const char *)strMonth(dt.month), 3);
        break;
      case 't':
        sprintf(helper, "%d", daysInMonth(dt.year, dt.month));
        strcat(buffer, (const char *)helper);
        break;

      // Year decoder
      case 'Y':
        sprintf(helper, "%d", dt.year);
        strcat(buffer, (const char *)helper);
        break;
      case 'y':
        sprintf(helper, "%02d", dt.year - 2000);
        strcat(buffer, (const char *)helper);
        break;
      case 'L':
        sprintf(helper, "%d", isLeapYear(dt.year));
        strcat(buffer, (const char *)helper);
        break;

      // Hour decoder
      case 'H':
        sprintf(helper, "%02d", dt.hour);
        strcat(buffer, (const char *)helper);
        break;
      case 'G':
        sprintf(helper, "%d", dt.hour);
        strcat(buffer, (const char *)helper);
        break;
      case 'h':
        sprintf(helper, "%02d", hour12(dt.hour));
        strcat(buffer, (const char *)helper);
        break;
      case 'g':
        sprintf(helper, "%d", hour12(dt.hour));
        strcat(buffer, (const char *)helper);
        break;
      case 'A':
        strcat(buffer, (const char *)strAmPm(dt.hour, true));
        break;
      case 'a':
        strcat(buffer, (const char *)strAmPm(dt.hour, false));
        break;

      // Minute decoder
      case 'i':
        sprintf(helper, "%02d", dt.minute);
        strcat(buffer, (const char *)helper);
        break;

      // Second decoder
      case 's':
        sprintf(helper, "%02d", dt.second);
        strcat(buffer, (const char *)helper);
        break;

      // Misc decoder
      case 'U':
        sprintf(helper, "%lu", dt.unixtime);
        strcat(buffer, (const char *)helper);
        break;

      default:
        strncat(buffer, dateFormat, 1);
        break;
    }
    dateFormat++;
  }

  return buffer;
}

char *DS3231::dateFormat(const char *dateFormat, RTCAlarmTime dt) {
  char buffer[255];

  buffer[0] = 0;

  char helper[11];

  while (*dateFormat != '\0') {
    switch (dateFormat[0]) {
      // Day decoder
      case 'd':
        sprintf(helper, "%02d", dt.day);
        strcat(buffer, (const char *)helper);
        break;
      case 'j':
        sprintf(helper, "%d", dt.day);
        strcat(buffer, (const char *)helper);
        break;
      case 'l':
        strcat(buffer, (const char *)strDayOfWeek(dt.day));
        break;
      case 'D':
        strncat(buffer, strDayOfWeek(dt.day), 3);
        break;
      case 'N':
        sprintf(helper, "%d", dt.day);
        strcat(buffer, (const char *)helper);
        break;
      case 'w':
        sprintf(helper, "%d", (dt.day + 7) % 7);
        strcat(buffer, (const char *)helper);
        break;
      case 'S':
        strcat(buffer, (const char *)strDaySufix(dt.day));
        break;

      // Hour decoder
      case 'H':
        sprintf(helper, "%02d", dt.hour);
        strcat(buffer, (const char *)helper);
        break;
      case 'G':
        sprintf(helper, "%d", dt.hour);
        strcat(buffer, (const char *)helper);
        break;
      case 'h':
        sprintf(helper, "%02d", hour12(dt.hour));
        strcat(buffer, (const char *)helper);
        break;
      case 'g':
        sprintf(helper, "%d", hour12(dt.hour));
        strcat(buffer, (const char *)helper);
        break;
      case 'A':
        strcat(buffer, (const char *)strAmPm(dt.hour, true));
        break;
      case 'a':
        strcat(buffer, (const char *)strAmPm(dt.hour, false));
        break;

      // Minute decoder
      case 'i':
        sprintf(helper, "%02d", dt.minute);
        strcat(buffer, (const char *)helper);
        break;

      // Second decoder
      case 's':
        sprintf(helper, "%02d", dt.second);
        strcat(buffer, (const char *)helper);
        break;

      default:
        strncat(buffer, dateFormat, 1);
        break;
    }
    dateFormat++;
  }

  return buffer;
}

RTCDateTime DS3231::getDateTime(void) {
  int values[7];

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_TIME);
#else
  Wire.send(DS3231_REG_TIME);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 7);

  while (!Wire.available()) {
  };

  for (int i = 6; i >= 0; i--) {
#if ARDUINO >= 100
    values[i] = bcd2dec(Wire.read());
#else
    values[i] = bcd2dec(Wire.receive());
#endif
  }

  Wire.endTransmission();

  t.year = values[0] + 2000;
  t.month = values[1];
  t.day = values[2];
  t.dayOfWeek = values[3];
  t.hour = values[4];
  t.minute = values[5];
  t.second = values[6];
  t.unixtime = unixtime();

  return t;
}

uint8_t DS3231::isReady(void) { return true; }

void DS3231::enableOutput(bool enabled) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  value &= 0b11111011;
  value |= (!enabled << 2);

  writeRegister8(DS3231_REG_CONTROL, value);
}

void DS3231::setBattery(bool timeBattery, bool squareBattery) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  if (squareBattery) {
    value |= 0b01000000;
  } else {
    value &= 0b10111111;
  }

  if (timeBattery) {
    value &= 0b01111011;
  } else {
    value |= 0b10000000;
  }

  writeRegister8(DS3231_REG_CONTROL, value);
}

bool DS3231::isOutput(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  value &= 0b00000100;
  value >>= 2;

  return !value;
}

void DS3231::setOutput(DS3231_sqw_t mode) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  value &= 0b11100111;
  value |= (mode << 3);

  writeRegister8(DS3231_REG_CONTROL, value);
}

DS3231_sqw_t DS3231::getOutput(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  value &= 0b00011000;
  value >>= 3;

  return (DS3231_sqw_t)value;
}

void DS3231::enable32kHz(bool enabled) {
  uint8_t value;

  value = readRegister8(DS3231_REG_STATUS);

  value &= 0b11110111;
  value |= (enabled << 3);

  writeRegister8(DS3231_REG_STATUS, value);
}

bool DS3231::is32kHz(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_STATUS);

  value &= 0b00001000;
  value >>= 3;

  return value;
}

void DS3231::forceConversion(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_CONTROL);

  value |= 0b00100000;

  writeRegister8(DS3231_REG_CONTROL, value);

  do {
  } while ((readRegister8(DS3231_REG_CONTROL) & 0b00100000) != 0);
}

float DS3231::readTemperature(void) {
  uint8_t msb, lsb;

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_TEMPERATURE);
#else
  Wire.send(DS3231_REG_TEMPERATURE);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 2);

  while (!Wire.available()) {
  };

#if ARDUINO >= 100
  msb = Wire.read();
  lsb = Wire.read();
#else
  msb = Wire.receive();
  lsb = Wire.receive();
#endif

  return ((((short)msb << 8) | (short)lsb) >> 6) / 4.0f;
}

RTCAlarmTime DS3231::getAlarm1(void) {
  uint8_t values[4];
  RTCAlarmTime a;

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_1);
#else
  Wire.send(DS3231_REG_ALARM_1);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 4);

  while (!Wire.available()) {
  };

  for (int i = 3; i >= 0; i--) {
#if ARDUINO >= 100
    values[i] = bcd2dec(Wire.read() & 0b01111111);
#else
    values[i] = bcd2dec(Wire.receive() & 0b01111111);
#endif
  }

  Wire.endTransmission();

  a.day = values[0];
  a.hour = values[1];
  a.minute = values[2];
  a.second = values[3];

  return a;
}

DS3231_alarm1_t DS3231::getAlarmType1(void) {
  uint8_t values[4];
  uint8_t mode = 0;

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_1);
#else
  Wire.send(DS3231_REG_ALARM_1);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 4);

  while (!Wire.available()) {
  };

  for (int i = 3; i >= 0; i--) {
#if ARDUINO >= 100
    values[i] = bcd2dec(Wire.read());
#else
    values[i] = bcd2dec(Wire.receive());
#endif
  }

  Wire.endTransmission();

  mode |= ((values[3] & 0b01000000) >> 6);
  mode |= ((values[2] & 0b01000000) >> 5);
  mode |= ((values[1] & 0b01000000) >> 4);
  mode |= ((values[0] & 0b01000000) >> 3);
  mode |= ((values[0] & 0b00100000) >> 1);

  return (DS3231_alarm1_t)mode;
}

void DS3231::setAlarm1(uint8_t dydw, uint8_t hour, uint8_t minute,
                       uint8_t second, DS3231_alarm1_t mode, bool armed) {
  second = dec2bcd(second);
  minute = dec2bcd(minute);
  hour = dec2bcd(hour);
  dydw = dec2bcd(dydw);

  switch (mode) {
    case DS3231_EVERY_SECOND:
      second |= 0b10000000;
      minute |= 0b10000000;
      hour |= 0b10000000;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_S:
      second &= 0b01111111;
      minute |= 0b10000000;
      hour |= 0b10000000;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_M_S:
      second &= 0b01111111;
      minute &= 0b01111111;
      hour |= 0b10000000;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_H_M_S:
      second &= 0b01111111;
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_DT_H_M_S:
      second &= 0b01111111;
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw &= 0b01111111;
      break;

    case DS3231_MATCH_DY_H_M_S:
      second &= 0b01111111;
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw &= 0b01111111;
      dydw |= 0b01000000;
      break;
  }

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_1);
  Wire.write(second);
  Wire.write(minute);
  Wire.write(hour);
  Wire.write(dydw);
#else
  Wire.send(DS3231_REG_ALARM_1);
  Wire.send(second);
  Wire.send(minute);
  Wire.send(hour);
  Wire.send(dydw);
#endif

  Wire.endTransmission();

  armAlarm1(armed);

  clearAlarm1();
}

bool DS3231::isAlarm1(bool clear) {
  uint8_t alarm;

  alarm = readRegister8(DS3231_REG_STATUS);
  alarm &= 0b00000001;

  if (alarm && clear) {
    clearAlarm1();
  }

  return alarm;
}

void DS3231::armAlarm1(bool armed) {
  uint8_t value;
  value = readRegister8(DS3231_REG_CONTROL);

  if (armed) {
    value |= 0b00000001;
  } else {
    value &= 0b11111110;
  }

  writeRegister8(DS3231_REG_CONTROL, value);
}

bool DS3231::isArmed1(void) {
  uint8_t value;
  value = readRegister8(DS3231_REG_CONTROL);
  value &= 0b00000001;
  return value;
}

void DS3231::clearAlarm1(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_STATUS);
  value &= 0b11111110;

  writeRegister8(DS3231_REG_STATUS, value);
}

RTCAlarmTime DS3231::getAlarm2(void) {
  uint8_t values[3];
  RTCAlarmTime a;

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_2);
#else
  Wire.send(DS3231_REG_ALARM_2);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 3);

  while (!Wire.available()) {
  };

  for (int i = 2; i >= 0; i--) {
#if ARDUINO >= 100
    values[i] = bcd2dec(Wire.read() & 0b01111111);
#else
    values[i] = bcd2dec(Wire.receive() & 0b01111111);
#endif
  }

  Wire.endTransmission();

  a.day = values[0];
  a.hour = values[1];
  a.minute = values[2];
  a.second = 0;

  return a;
}

DS3231_alarm2_t DS3231::getAlarmType2(void) {
  uint8_t values[3];
  uint8_t mode = 0;

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_2);
#else
  Wire.send(DS3231_REG_ALARM_2);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 3);

  while (!Wire.available()) {
  };

  for (int i = 2; i >= 0; i--) {
#if ARDUINO >= 100
    values[i] = bcd2dec(Wire.read());
#else
    values[i] = bcd2dec(Wire.receive());
#endif
  }

  Wire.endTransmission();

  mode |= ((values[2] & 0b01000000) >> 5);
  mode |= ((values[1] & 0b01000000) >> 4);
  mode |= ((values[0] & 0b01000000) >> 3);
  mode |= ((values[0] & 0b00100000) >> 1);

  return (DS3231_alarm2_t)mode;
}

void DS3231::setAlarm2(uint8_t dydw, uint8_t hour, uint8_t minute,
                       DS3231_alarm2_t mode, bool armed) {
  minute = dec2bcd(minute);
  hour = dec2bcd(hour);
  dydw = dec2bcd(dydw);

  switch (mode) {
    case DS3231_EVERY_MINUTE:
      minute |= 0b10000000;
      hour |= 0b10000000;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_M:
      minute &= 0b01111111;
      hour |= 0b10000000;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_H_M:
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw |= 0b10000000;
      break;

    case DS3231_MATCH_DT_H_M:
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw &= 0b01111111;
      break;

    case DS3231_MATCH_DY_H_M:
      minute &= 0b01111111;
      hour &= 0b01111111;
      dydw &= 0b01111111;
      dydw |= 0b01000000;
      break;
  }

  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(DS3231_REG_ALARM_2);
  Wire.write(minute);
  Wire.write(hour);
  Wire.write(dydw);
#else
  Wire.send(DS3231_REG_ALARM_2);
  Wire.send(minute);
  Wire.send(hour);
  Wire.send(dydw);
#endif

  Wire.endTransmission();

  armAlarm2(armed);

  clearAlarm2();
}

void DS3231::armAlarm2(bool armed) {
  uint8_t value;
  value = readRegister8(DS3231_REG_CONTROL);

  if (armed) {
    value |= 0b00000010;
  } else {
    value &= 0b11111101;
  }

  writeRegister8(DS3231_REG_CONTROL, value);
}

bool DS3231::isArmed2(void) {
  uint8_t value;
  value = readRegister8(DS3231_REG_CONTROL);
  value &= 0b00000010;
  value >>= 1;
  return value;
}

void DS3231::clearAlarm2(void) {
  uint8_t value;

  value = readRegister8(DS3231_REG_STATUS);
  value &= 0b11111101;

  writeRegister8(DS3231_REG_STATUS, value);
}

bool DS3231::isAlarm2(bool clear) {
  uint8_t alarm;

  alarm = readRegister8(DS3231_REG_STATUS);
  alarm &= 0b00000010;

  if (alarm && clear) {
    clearAlarm2();
  }

  return alarm;
}

uint8_t DS3231::bcd2dec(uint8_t bcd) { return ((bcd / 16) * 10) + (bcd % 16); }

uint8_t DS3231::dec2bcd(uint8_t dec) { return ((dec / 10) * 16) + (dec % 10); }

char *DS3231::strDayOfWeek(uint8_t dayOfWeek) {
  switch (dayOfWeek) {
    case 1:
      return "Monday";
      break;
    case 2:
      return "Tuesday";
      break;
    case 3:
      return "Wednesday";
      break;
    case 4:
      return "Thursday";
      break;
    case 5:
      return "Friday";
      break;
    case 6:
      return "Saturday";
      break;
    case 7:
      return "Sunday";
      break;
    default:
      return "Unknown";
  }
}

char *DS3231::strMonth(uint8_t month) {
  switch (month) {
    case 1:
      return "January";
      break;
    case 2:
      return "February";
      break;
    case 3:
      return "March";
      break;
    case 4:
      return "April";
      break;
    case 5:
      return "May";
      break;
    case 6:
      return "June";
      break;
    case 7:
      return "July";
      break;
    case 8:
      return "August";
      break;
    case 9:
      return "September";
      break;
    case 10:
      return "October";
      break;
    case 11:
      return "November";
      break;
    case 12:
      return "December";
      break;
    default:
      return "Unknown";
  }
}

char *DS3231::strAmPm(uint8_t hour, bool uppercase) {
  if (hour < 12) {
    if (uppercase) {
      return "AM";
    } else {
      return "am";
    }
  } else {
    if (uppercase) {
      return "PM";
    } else {
      return "pm";
    }
  }
}

char *DS3231::strDaySufix(uint8_t day) {
  if (day % 10 == 1) {
    return "st";
  } else if (day % 10 == 2) {
    return "nd";
  }
  if (day % 10 == 3) {
    return "rd";
  }

  return "th";
}

uint8_t DS3231::hour12(uint8_t hour24) {
  if (hour24 == 0) {
    return 12;
  }

  if (hour24 > 12) {
    return (hour24 - 12);
  }

  return hour24;
}

long DS3231::time2long(uint16_t days, uint8_t hours, uint8_t minutes,
                       uint8_t seconds) {
  return ((days * 24L + hours) * 60 + minutes) * 60 + seconds;
}

uint16_t DS3231::dayInYear(uint16_t year, uint8_t month, uint8_t day) {
  uint16_t fromDate;
  uint16_t toDate;

  fromDate = date2days(year, 1, 1);
  toDate = date2days(year, month, day);

  return (toDate - fromDate);
}

bool DS3231::isLeapYear(uint16_t year) { return (year % 4 == 0); }

uint8_t DS3231::daysInMonth(uint16_t year, uint8_t month) {
  uint8_t days;

  days = pgm_read_byte(daysArray + month - 1);

  if ((month == 2) && isLeapYear(year)) {
    ++days;
  }

  return days;
}

uint16_t DS3231::date2days(uint16_t year, uint8_t month, uint8_t day) {
  year = year - 2000;

  uint16_t days16 = day;

  for (uint8_t i = 1; i < month; ++i) {
    days16 += pgm_read_byte(daysArray + i - 1);
  }

  if ((month == 2) && isLeapYear(year)) {
    ++days16;
  }

  return days16 + 365 * year + (year + 3) / 4 - 1;
}

uint32_t DS3231::unixtime(void) {
  uint32_t u;

  u = time2long(date2days(t.year, t.month, t.day), t.hour, t.minute, t.second);
  u += 946681200;

  return u;
}

uint8_t DS3231::conv2d(const char *p) {
  uint8_t v = 0;

  if ('0' <= *p && *p <= '9') {
    v = *p - '0';
  }

  return 10 * v + *++p - '0';
}

uint8_t DS3231::dow(uint16_t y, uint8_t m, uint8_t d) {
  uint8_t dow;

  y -= m < 3;
  dow =
      ((y + y / 4 - y / 100 + y / 400 + pgm_read_byte(dowArray + (m - 1)) + d) %
       7);

  if (dow == 0) {
    return 7;
  }

  return dow;
}

void DS3231::writeRegister8(uint8_t reg, uint8_t value) {
  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(reg);
  Wire.write(value);
#else
  Wire.send(reg);
  Wire.send(value);
#endif
  Wire.endTransmission();
}

uint8_t DS3231::readRegister8(uint8_t reg) {
  uint8_t value;
  Wire.beginTransmission(DS3231_ADDRESS);
#if ARDUINO >= 100
  Wire.write(reg);
#else
  Wire.send(reg);
#endif
  Wire.endTransmission();

  Wire.requestFrom(DS3231_ADDRESS, 1);
  while (!Wire.available()) {
  };
#if ARDUINO >= 100
  value = Wire.read();
#else
  value = Wire.receive();
#endif;
  Wire.endTransmission();

  return value;
}
